MAINT: Do not use deprecated ``mktemp()``

lyzlisa opened the pull request:
Closes #12071 

* Changed use of `mktemp` to `mkstemp` since `mktemp` was deprecated
* No change required for `numpy/lib/tests/test_format.py`
* Coverage report: 84%

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

lyzlisa committed:
MAINT: Deprecate `mktemp()` usage in tests

Co-authored-by: alize-papp <68250865+alize-papp@users.noreply.github.com>

rossbar commented:
One difference between `mktemp` and `mkstemp` is that the latter creates the file handle, which the caller is [then responsible for closing](https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp). I'm not sure how the files are handled in the modified tests, but IMO it's worth double-checking that things are explicitly closed.

lyzlisa commented:
In the original code when `mktemp()` was used, `fp` was deleted at the end of the test.

With the use of `mkstemp()`, `fp` is still being deleted at the end of the test.

Would it be safer to use a context manager to explicitly close the file before deleting it?

mattip commented:
These days I think a [pytest `tmp_path`](https://docs.pytest.org/en/latest/how-to/tmpdir.html?highlight=temporary#the-tmp-path-fixture) would be cleaner. The test signature would become `test_something(self, tmp_path)` and pytest magically fills that in with a temporary pathlib object.

lyzlisa committed:
MAINT: Use `tmp_path` fixture instead of `mkstemp()` function

Co-authored-by: alize-papp <68250865+alize-papp@users.noreply.github.com>

alize-papp commented:
We used pytest `tmp_path` in `test_memmap.py` as suggested.
In `test_multiarray.py`  it feels like we don't need to, since there is: 
```
    def teardown(self):
    shutil.rmtree(self.tempdir)
```
that already deletes everything. Does that make sense?

rgommers commented:
> that already deletes everything. Does that make sense?

That sounds right. It's not impossible that a `ResourceWarning` will still be emitted because of discarding the open file handle, in that case CI will fail.

mattip commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -4818,7 +4818,7 @@ def setup(self):
         self.dtype = self.x.dtype
         self.tempdir = tempfile.mkdtemp()
-        self.filename = tempfile.mktemp(dir=self.tempdir)
+        self.filename = tempfile.mkstemp(dir=self.tempdir)[1]
```
As [documented](https://docs.python.org/3/library/tempfile.html#tempfile.mkstemp)
> `mkstemp()` returns a tuple containing an OS-level handle to an open file (as would be returned by `os.open()`) and the absolute pathname of that file, in that order.

So someone needs to close that open handle. I would prefer to decouple all the tests that use `self.filename` by converting them to tests that use pytest's `tmp_file` instead. Then each test would get a new filename: one failing test would not influence another.

lyzlisa committed:
MAINT: Refactor test_multiarray.py to use pytest-native fixture

- `test_roundtrip_filename` and `test_roundtrip_pathlib` are now combined into a single test. The fixture covers both the filename as a string and a pathlib object
- Added fixture for testing localized decimal point to replace `test_locale`

mattip commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -4852,38 +4861,32 @@ def test_fromstring_count0(self):
         assert d.shape == (0,)
 
-    def test_empty_files_binary(self):
+    def test_empty_files_text(self, tmp_filename):
         with open(self.filename, 'w') as f:
             pass
         y = np.fromfile(self.filename)
         assert_(y.size == 0, "Array not empty")
```
I would suggest a deeper refactoring, using pytest's `tmp_file` with no extra fixtures needed in the class setup or teardown. Doesn't this work:

```suggestion
    def test_empty_files_text(self, tmp_file):
        with open(tmp_file, 'w') as f:
            pass
        y = np.fromfile(tmp_file)
        assert_(y.size == 0, "Array not empty")
```

mboutet replied:
@mattip, unless I'm mistaken, `tmp_file` is not a built-in pytest fixture. You can try by running this dummy test:

```python
def test_foobar(tmp_file):
    pass
```

It will fail because the fixture does not exist.

Besides, if such built-in fixture existed, it would probably also create the file on disk which is something we don't want for the tests. We only want a filename without pre-creating the file.

mattip replied:
Ahh, my bad. It is `tmp_path` not `tmp_file` which gives you a PathLib object pointing to a directory like `/tmp/pytest-of-matti/pytest-8/test_foobar0` where each invocation of pytest will increment the `pytest-8` part and delete old ones according to [some logic](https://docs.pytest.org/en/latest/how-to/tmpdir.html#the-default-base-temporary-directory), and `test_foobar0` is the name of the test with a 0 for good measure. So this test would be
```
def test_empty_files_text(self, tmp_path):
        with open(tmp_path / 'empty', 'w') as f:
            pass
        y = np.fromfile(tmp_path / 'empty')
        assert_(y.size == 0, "Array not empty")
```

mboutet replied:
Ok, but if `tmp_path` is used directly, then all references to `self.filename` must be removed. Another benefit of the custom fixture `tmp_filename` is that it is parametrized to cover two cases with very low effort (`filename` is a `str` and `filename` is a path-like object). Doing so allows to combine at least two tests into one test.

lyzlisa committed:
MAINT: Remove outdated use of `mkdtemp()` in test_memmap.py

mattip commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -4811,17 +4810,27 @@ def test_invalid_axis(self): # gh-7528
-        shutil.rmtree(self.tempdir)
+        yield
+        delattr(self, "dtype")
+        delattr(self, "x")
```
I think `setup/teardown` is a better paradigm here than using a `yield` like this. Why do you want to re-create `dtype` and `x`?

mattip replied:
In this case then I think it would be clearer if `x` was passed in to the tests that need it as an additional argument, rather than stuffing it onto `self`. Tests that need `dtype` should just call `x.dtype`. As I commented elsewhere, this use/abuse of `self.x` seems weird.

lyzlisa replied:
I refactored `TestIO` to use a fixture (`x`) and also updated all the `self.filename` references to use the value from the `tmp_filename` fixture instead, which remove all attributes stored in `self`.

Is it too much out of scope to refactor `TestMemmap` to use fixtures too since the original issue was to remove the deprecated `mktemp()` function?

mattip replied:
I think the `TestMemmap` is fine as is. I would like to keep the use of fixtures to a minimum, it adds a layer of indirection that is harder to parse.

lyzlisa committed:
MAINT: Refactor TestIO setup into pytest-fixture

- Removed `x`, `dtype`, and `filename` from class attributes

lyzlisa committed:
MAINT: Fix typos in `test_read_shorter_than_count_subarray()` comment

mattip commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -5057,46 +5060,61 @@ def dup_bigint(fd):
+        else:
+            yield
+
+    def test_nan(self, tmp_filename, comma_decimal_point_locale):
```
It is not clear form the code how comma_decimal_point_locale is ever used. A comment would be helpful.

mattip replied:
just to be clear: the previous code was also opaque as to what this does

mattip commented:
Thanks, I think this is a big improvement.

mattip reviewed (approved):
LGTM, besides one nit for a comment. Would anyone else like to take a look?

lyzlisa committed:
MAINT: Refactor `comma_decimal_point_locale` to be more explicit

- Renamed `comma_decimal_point_locale` to `decimal_sep_localization`
- Added a docstring explaining its usage

rossbar commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -4811,17 +4810,23 @@ def test_invalid_axis(self): # gh-7528
     """Test tofile, fromfile, tobytes, and fromstring"""
 
-    def setup(self):
+    @pytest.fixture()
+    def x(self):
```
I was confused as to why this fixture had a parameter, but I guess that's because it's defined within a class. I've never seen that pattern used before. Alternatively this fixture could be defined outside of the class which would get rid of the confusing (and AFAICT unnecessary) parameter. Anyways, it's fine as-is I just wanted to bring it up in case this was intentional and I'd learn something new about fixtures :)

lyzlisa replied:
You are correct, and this fixture could be brought outside the `TestIO` class. However, because it is only used within `TestIO`, it made more sense to declare it under `TestIO`. If `x` were to be used in other tests outside `TestIO`, then it could be defined at the module-level, which is definitely one of the benefits of using fixtures: easy reusability.

To get rid of the `self` parameter, I guess we can write:
```
    @staticmethod
    @pytest.fixture()
    def x():
        ...
```

rossbar replied:
No worries, I was just curious - it's fine as-is.

rossbar commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -5057,46 +5060,81 @@ def dup_bigint(fd):
         assert_array_equal(y, value)
 
-    def test_nan(self):
+    @pytest.fixture(params=["period", "comma"])
+    def decimal_sep_localization(self, request):
```
This is a really nifty way of doing this!

rossbar reviewed (approved):
LGTM as well - using pytest features for the tmp file handling is a nice improvement IMO. Thanks @lyzlisa !
